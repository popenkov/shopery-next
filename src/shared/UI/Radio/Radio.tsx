'use client';

import { ChangeEvent, InputHTMLAttributes, memo, useRef } from 'react';

import cn from 'classnames';


import cls from './Radio.module.scss';

type HTMLRadioProps = Omit<
    InputHTMLAttributes<HTMLInputElement>,
    'value' | 'onChange' | 'readOnly'
>;

interface RadioProps extends HTMLRadioProps {
    className?: string;
    name: string;
    value: string | number;
    text?: string;
    amount?: number;
    onChange?: (value: string) => void;
}

export const Radio = memo((props: RadioProps) => {
    const { className, value, onChange, text, name, amount, ...otherProps } =
        props;
    const ref = useRef<HTMLInputElement>(null);

    const handleRadioChange = (evt: ChangeEvent<HTMLInputElement>) => {
        onChange?.(evt.target.value);
    };

    return (
        <label className={cn(cls.radio, className)}>
            <input
                ref={ref}
                type="radio"
                name={name}
                value={value}
                onChange={handleRadioChange}
                className={cls.input}
                {...otherProps}
            />
            <span className={cls.fakeInput} />
            {text && <span className={cls.placeholder}>{text}</span>}
            {typeof amount === 'number' && (
                <span className={cls.amount}>( ${amount} )</span>
            )}
        </label>
    );
});

Radio.displayName = 'Radio';
