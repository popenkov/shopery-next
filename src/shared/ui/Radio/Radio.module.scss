@import '../../styles/mixins/flexbox';
@import '../../styles/mixins/box';
@import '../../styles/mixins/text-styles';
@import '../../styles/mixins/absolute-positioning';

.input {
  position: absolute;
  width: 16px;
  height: 16px;
  opacity: 0;
}

.fakeInput {
  @include flex-center;
  @include box(20px);

  position: relative;
  border: 1px solid var(--gray-200);
  border-radius: 50%;
  cursor: pointer;

  &::after {
    @include absolute-center;
    @include box(12px);

    border-radius: 50%;
    transition: background 0.125s ease;
    content: '';
  }
}

.input:checked + .fakeInput {
  border-color: var(--primary);
  border-width: 2px;
}

.input:checked + .fakeInput::after {
  background: var(--primary);
}

.label {
  @include body-s;

  cursor: pointer;
  user-select: none;
}

.amount {
  @include body-s;

  color: var(--gray-500);
}

.label + .amount {
  margin-left: -4px;
}

.radio {
  position: relative;
  display: flex;
  gap: 8px;
  justify-self: start;
  transition: var(--transition);

  @media (hover: hover) {
    &:hover {
      color: var(--primary);

      .fakeInput {
        border-color: var(--primary);
      }
    }
  }
}
