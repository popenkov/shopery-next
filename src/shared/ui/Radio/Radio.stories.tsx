import type { Meta, StoryObj } from '@storybook/react';

import { Radio } from './Radio';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'shared/Radio',
  component: Radio,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Radio>;

export default meta;
type Story = StoryObj<typeof meta>;

export const RadioComponent: Story = {
  args: {
    name: 'name',
    value: 'value',
    text: 'test',
    amount: 33,
  },
};
