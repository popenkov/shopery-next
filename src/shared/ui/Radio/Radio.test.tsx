import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Radio } from './Radio';

describe('Radio component', () => {
  it('renders correctly', () => {
    const { container } = render(<Radio name="test" value="test-value" />);
    expect(container).toMatchSnapshot();
  });

  it('calls onChange callback when radio button is changed', () => {
    const onChange = jest.fn();
    const { getByRole } = render(<Radio name="test" value="test-value" onChange={onChange} />);
    const radioButton = getByRole('radio');
    fireEvent.click(radioButton);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('test-value');
  });

  it('renders text and amount correctly', () => {
    const { getByText } = render(
      <Radio name="test" value="test-value" text="Test text" amount={10} />,
    );
    expect(getByText('Test text')).toBeInTheDocument();
    expect(getByText('(10)')).toBeInTheDocument();
  });

  it('sets checked state correctly', () => {
    const { getByRole } = render(<Radio name="test" value="test-value" checked />);
    const radioButton = getByRole('radio');
    expect(radioButton).toBeChecked();
  });

  it('does not call onChange callback when radio button is not changed', () => {
    const onChange = jest.fn();
    const { getByRole } = render(<Radio name="test" value="test-value" onChange={onChange} />);
    const radioButton = getByRole('radio');
    fireEvent.click(radioButton);
    fireEvent.click(radioButton);
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
