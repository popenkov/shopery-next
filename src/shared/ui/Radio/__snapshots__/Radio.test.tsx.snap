// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component renders correctly 1`] = `
<div>
  <label
    class="radio"
  >
    <input
      class="input"
      name="test"
      type="radio"
      value="test-value"
    />
    <span
      class="fakeInput"
    />
  </label>
</div>
`;
